File: src/0078_subsets.js
function subsets(nums) {
    const res = [];
    const subset = [];
    const dfs = (i) => {
        if (i >= nums.length) {
            res.push(subset.slice());
            return;
        }
        subset.push(nums[i]);
        dfs(i + 1);
        subset.pop();
        dfs(i + 1);
    };
    dfs(0);
    return res;
}

module.exports = { subsets };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(subsets([1, 2, 3]), [[1, 2, 3], [1, 2], [1, 3], [1], [2, 3], [2], [3], []]);
}

File: src/0078_subsets.rs
struct Solution;

impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut res = Vec::new();
        let mut subset = Vec::new();
        dfs(0, &nums, &mut subset, &mut res);
        res
    }
}

fn dfs(index: usize, nums: &[i32], subset: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {
    if index >= nums.len() {
        res.push(subset.clone());
        return;
    }

    subset.push(nums[index]);
    dfs(index + 1, nums, subset, res);
    subset.pop();
    dfs(index + 1, nums, subset, res);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::subsets(vec![1, 2, 3]),
            vec![
                vec![1, 2, 3],
                vec![1, 2],
                vec![1, 3],
                vec![1],
                vec![2, 3],
                vec![2],
                vec![3],
                vec![]
            ]
        );
    }
}
